File: src/test/java/komu/blunt/eval/EvaluatorPerformanceTest.java
package komu.blunt.eval;

import komu.blunt.parser.Parser;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestWatcher;
import org.junit.runner.Description;

import java.io.IOException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static java.lang.String.format;
import static org.junit.Assert.fail;

public class EvaluatorPerformanceTest {

    private static final Evaluator evaluator = new Evaluator();

    @Rule
    public final Profiler profiler = new Profiler();
    
    @BeforeClass
    public static void loadPrelude() throws IOException {
        evaluator.loadResource("prelude.blunt");
    }
    
    @Test
    @MaxSteps(2_021_216)
    public void sorting() {
        evaluate("sort (range 0 500)");
    }

    private static Object evaluate(String expr) {
        return evaluator.evaluate(Parser.$classobj.parseExpression(expr));
    }

    private static class Profiler extends TestWatcher {

        private long startTime;
        private long startSteps;
        
        @Override
        protected void starting(Description description) {
            startTime = System.currentTimeMillis();
            startSteps = evaluator.getSteps();
        }

        @Override
        protected void finished(Description description) {
            long elapsedTime = System.currentTimeMillis() - startTime;
            long elapsedSteps = evaluator.getSteps() - startSteps;

            System.out.printf("%s: %,d steps in %dms\n", description.getMethodName(), elapsedSteps, elapsedTime);

            MaxSteps maxSteps = description.getAnnotation(MaxSteps.class);
            if (maxSteps != null && elapsedSteps > maxSteps.value())
                fail(format("expected maximum %,d steps, but got %,d", maxSteps.value(), elapsedSteps));
        }
    }

    @Retention(RetentionPolicy.RUNTIME)
    private @interface MaxSteps {
        long value();
    }
}

File: src/test/kotlin/komu/blunt/eval/EvaluatorPerformanceTest.kt
package komu.blunt.eval;

import komu.blunt.parser.parseExpression
import org.junit.Before
import org.junit.Test
import kotlin.system.measureTimeMillis
import kotlin.test.fail

class EvaluatorPerformanceTest {

    val evaluator = Evaluator()

    @Before
    fun loadPrelude() {
        evaluator.loadResource("prelude.blunt");
    }
    
    @Test
    fun sorting() {
        evaluate("sort (range 0 500)", maxSteps = 2021216)
    }

    fun evaluate(expr: String, maxSteps: Long) {
        val startSteps = evaluator.steps
        val total = measureTimeMillis {
            evaluator.evaluate(parseExpression(expr))
        }
        val steps = evaluator.steps - startSteps

        print("evaluation of '$expr' took $steps steps in ${total}ms")

        if (steps > maxSteps)
            fail("expected maximum $maxSteps steps, but got $steps")
    }
}
